### PR TITLE
Merge timeline plus

### PR DESCRIPTION
**Pull request by yotamberk**
This PR should merge all the work done in https://github.com/yotamberk/timeline-plus in to this repo.
Unfortunately I could not preserve the amazing original contributors of some of the features implemented.

### CODE REVIEW

**Comment by mojoaxel** [comment]
I'll try to review and test it this week.
@yotamberg Could you please preparerelease nodes e.g  in the HISTORY.md file. Do you think it is possible to list the original pull requests and contributors?

**Comment by yotamberk** [comment]
Will do. I'll get the HISTORY.md prepared once I get this branch up and running again

**Comment by mojoaxel** [comment]
@yotamberk By the way: It looks like you can now create pull-request DRAFTS for work-in-progress. No idea how this works but here is an example: https://github.com/visjs/vis-data/pull/4

File: HISTORY.md [hunk 1]
```diff
@@ -2,6 +2,51 @@
 
 http://visjs.org
 
+## 2019-07-19, version 6.0.0
```
Comment (mojoaxel) [inline HISTORY.md]: The HISTORY is perfect! Good job!

**Comment by mojoaxel** [comment]
@yotamberk I'm really looking forward to merge this. What is the state of this PR?

**Comment by yotamberk** [comment]
I have a few clean ups I have to do. I believe I can get it done by tomorrow.

**Comment by yotamberk** [comment]
Can't do it today. Will have to wait for tomorrow. Sorry...

**Comment by yotamberk** [comment]
Egh.... Many examples seems to be broken... I'm working as fast as I can to try to fix the problems causing the bugs.

File: rollup.config.js [hunk 2]
```diff
@@ -38,16 +37,15 @@ export default [{
 		exports: 'default',
 		format: 'umd',
 		globals: GLOBALS,
-		sourceMap: true
+		sourcemap: true
```
Comment (mojoaxel) [inline rollup.config.js]: I had problems with sourcemap combined with banner. Does this work?

**Comment by yotamberk** [comment]
FIXED! @mojoaxel This can be reviewed now.
This won't be easy to review...

File: lib/timeline/component/ItemSet.js [hunk 3]
```diff
@@ -420,45 +416,50 @@ class ItemSet extends Component {
...
         if (typeof options.groupEditable === 'boolean') {
+          this.options.groupEditable = {};
           this.options.groupEditable.order  = options.groupEditable;
           this.options.groupEditable.add    = options.groupEditable;
           this.options.groupEditable.remove = options.groupEditable;
         }
         else if (typeof options.groupEditable === 'object') {
-          util.selectiveExtend(['order', 'add', 'remove'], this.options.groupEditable, options.groupEditable);
+          this.options.groupEditable = { ...this.options.groupEditable, ...options.groupEditable }
+          util.selectiveExten(['order', 'add', 'remove'], this.options.groupEditable, options.groupEditable);
```
Comment (mojoaxel) [inline lib/timeline/component/ItemSet.js]: Something went wrong here

File: lib/timeline/Core.js [hunk 4]
```diff
@@ -392,13 +392,7 @@ class Core {
...
   setOptions(options) {
     if (options) {
-      // copy the known options
-      const fields = [
-        'width', 'height', 'minHeight', 'maxHeight', 'autoResize',
-        'start', 'end', 'clickToUse', 'dataAttributes', 'hiddenDates',
-        'locale', 'locales', 'moment', 'preferZoom', 'rtl', 'zoomKey', 'horizontalScroll', 'verticalScroll'
-      ];
-      util.selectiveExtend(fields, this.options, options); 
+      this.options = { ...this.options, ...options }
```
Comment (mojoaxel) [inline lib/timeline/Core.js]: I think we should keep the `selectiveExtend` it ensures that no other options get overwritten (e.g. by mistake).

**Comment by yotamberk** [comment]
@mojoaxel Done! Finally...
I've managed to fix everything and now everything seems to be working again.
Please merge this so we can continue on with this branch and the work on timeline here!

File: rollup.config.js [hunk 5]
```diff
@@ -34,7 +34,7 @@ export default [{
 	output: {
 		file: 'dist/vis-timeline-graph2d.min.js',
 		name: 'vis',
-		exports: 'named',
```
Comment (mojoaxel) [inline rollup.config.js]: I think we need `named` here!

File: lib/timeline/component/item/Item.js [hunk 6]
```diff
@@ -13,508 +17,528 @@ import '../css/item.css';
...
+        me.parent.itemSet._onUpdateItem(me);
+        me.parent.itemSet.body.emitter.emit('doubleClick', {
+          event,
+          item: me.id
+        });
       });
-    });
+      this.hammer.on('panstart', me.parent.itemSet._onDragStart.bind(me.parent.itemSet));
+      this.hammer.on('panmove',  me.parent.itemSet._onDrag.bind(me.parent.itemSet));
+      this.hammer.on('panend',   me.parent.itemSet._onDragEnd.bind(me.parent.itemSet));
```
Comment (jczacharia) [inline lib/timeline/component/item/Item.js]: These should be `this.hammerDragCenter.*`
this.hammer does not exist.

Comment (jczacharia) [reply lib/timeline/component/item/Item.js]: @mojoaxel This should fix the inability to edit items.

File: lib/timeline/component/css/timeline.js [hunk 7]
```diff
@@ -1,4 +1,4 @@
-//import { StyleSheet } from 'react-native';
+import { StyleSheet } from 'react-native';
```
[the remark this answers is not in the document]
Comment (yotamberk) [reply lib/timeline/component/css/timeline.js]: Mistake

File: lib/timeline/component/item/Item.js (hunk 6 above)
Comment (yotamberk) [reply lib/timeline/component/item/Item.js]: Added

File: rollup.config.js (hunk 5 above)
Comment (yotamberk) [reply rollup.config.js]: Nope. We changed the `index.js` to `export default` right here:
https://github.com/visjs/vis-timeline/pull/5/files#diff-168726dbe96b3ce427e7fedce31bb0bcR43

**Comment by yotamberk** [comment]
> ## Editing items
> * The item editing seams broken
>   
>   * [customSnappingOfItems](https://visjs.github.io/vis-timeline/examples/timeline/editing/customSnappingOfItems.html)
>   * [editingItems](https://visjs.github.io/vis-timeline/examples/timeline/editing/editingItems.html)
> 
> e.g. the "delete" button (red cross) is missing and items do not move when dragged.
> 
> ## Stylling
> * The language switcher does not work:
>   https://visjs.github.io/vis-timeline/examples/timeline/styling/weekStyling.html

Fixed now after last commit

**Comment by yotamberk** [comment]
> ❤️ Looks really good!
> 
> * Please remove the file: examples/timeline/resources/img/Thumbs.db
> * Most Graph2d examples are broken and need to be fixed.
> * Please fix linting by calling `npm run lint`
> 
> I just noticed that the v6 release would be an good chance to introduce

Thanks!
* removed examples/timeline/resources/img/Thumbs.db
* I'm not intending to fix the examples if it's not a simple fix. I'm not maintaining this module. I would like to export it to a different repository ASAP.
* I'll start doing it now

**Comment by mojoaxel** [comment]
> I'm not intending to fix the examples if it's not a simple fix. I'm not maintaining this module. I would like to export it to a different repository ASAP.

I looked at it and graph2d is just a very small part of the whole timeline. It makes no sense to seperate it in my few. We should just keep it functional without giving it much thought.

**Comment by yotamberk** [comment]
Weird. When I run it, no warnings appear. 
But the VSCode does show the problems only after I open the file once.
Why would the eslint not show the same warnings?

**Comment by yotamberk** [comment]
@mojoaxel TBH, I think we should get rid of JSdoc. It seems to be quite deprecated and not used at all. I don;t think there is a reason to keep it. At least if not completely remove it, not make it required for lint. It's really not very useful.
In any case, I'd like to put this PR behind us and get this merged so we can continue with the work

**Comment by mojoaxel** [comment]
I'll have a look tomorrow. We can deactivate the jsdoc rule for linting.
I have one last look tomorrow and than we can merge, test and release :+1:

**Comment by mojoaxel** [comment]
I still see a few bugs. I think we should merge than fix the bugs and THEN release the v6. Ok?

**Comment by yotamberk** [comment]
YAY! It's done!

**Pull request merged**